sycl/include/sycl/ext/oneapi/experimental/cuda/non_uniform_algorithms.hpp

```diff
@@ -0,0 +1,370 @@
+                     IsBitXOR<T, BinaryOperation>::value,
+                 T>
+masked_reduction_cuda_sm80(Group g, T x, BinaryOperation binary_op,
+                           const uint32_t MemberMask) {
+  return __nvvm_redux_sync_xor(x, MemberMask);
+}
+////
+
+//// Shuffle based masked reduction impls
+
+// Cluster group reduction using shfls, T = double
+template <typename Group, typename T, class BinaryOperation>
+inline __SYCL_ALWAYS_INLINE
+    std::enable_if_t<ext::oneapi::experimental::is_fixed_size_group_v<Group> &&
+                         std::is_same_v<T, double>,
+                     T>
+    masked_reduction_cuda_shfls(Group g, T x, BinaryOperation binary_op,
+                                const uint32_t MemberMask) {
+  for (int i = g.get_local_range()[0] / 2; i > 0; i /= 2) {
+    int x_a, x_b;
+    asm volatile("mov.b64 {%0,%1},%2; \n\t" : "=r"(x_a), "=r"(x_b) : "l"(x));
+
+    auto tmp_a = __nvvm_shfl_sync_bfly_i32(MemberMask, x_a, -1, i);
+    auto tmp_b = __nvvm_shfl_sync_bfly_i32(MemberMask, x_b, -1, i);
+    double tmp;
+    asm volatile("mov.b64 %0,{%1,%2}; \n\t"
+                 : "=l"(tmp)
+                 : "r"(tmp_a), "r"(tmp_b));
+    x = binary_op(x, tmp);
+  }
+
+  return x;
+}
+
+// Cluster group reduction using shfls, T = float
+template <typename Group, typename T, class BinaryOperation>
+inline __SYCL_ALWAYS_INLINE
+    std::enable_if_t<ext::oneapi::experimental::is_fixed_size_group_v<Group> &&
+                         std::is_same_v<T, float>,
+                     T>
+    masked_reduction_cuda_shfls(Group g, T x, BinaryOperation binary_op,
+                                const uint32_t MemberMask) {
+
+  for (int i = g.get_local_range()[0] / 2; i > 0; i /= 2) {
+    auto tmp =
+        __nvvm_shfl_sync_bfly_i32(MemberMask, __nvvm_bitcast_f2i(x), -1, i);
+    x = binary_op(x, __nvvm_bitcast_i2f(tmp));
+  }
+  return x;
+}
+
+// Cluster group reduction using shfls, std::is_integral_v<T>
+template <typename Group, typename T, class BinaryOperation>
+inline __SYCL_ALWAYS_INLINE
+    std::enable_if_t<ext::oneapi::experimental::is_fixed_size_group_v<Group> &&
+                         std::is_integral_v<T>,
+                     T>
+    masked_reduction_cuda_shfls(Group g, T x, BinaryOperation binary_op,
+                                const uint32_t MemberMask) {
+
+  for (int i = g.get_local_range()[0] / 2; i > 0; i /= 2) {
+    auto tmp = __nvvm_shfl_sync_bfly_i32(MemberMask, x, -1, i);
+    x = binary_op(x, tmp);
+  }
+  return x;
+}
+
+template <typename Group, typename T>
+inline __SYCL_ALWAYS_INLINE std::enable_if_t<
+    ext::oneapi::experimental::is_user_constructed_group_v<Group>, T>
+non_uniform_shfl_T(const uint32_t MemberMask, T x, int delta) {
+  if constexpr (ext::oneapi::experimental::is_fixed_size_group_v<Group>) {
+    return __nvvm_shfl_sync_up_i32(MemberMask, x, delta, 0);
+  } else {
+    return __nvvm_shfl_sync_idx_i32(MemberMask, x, delta, 31);
```

Comment:
Something about this line looks off.  If it's an indexed shuffle, should it be accessing the current lane + delta?